doc/acl/Everflow-test-plan.md

```diff
@@ -0,0 +1,1134 @@
+```
+$ redis-cli -n 6 hgetall 'SWITCH_CAPABILITY|switch'
+ 1) "MIRROR"
+ 2) "true"
+ 3) "MIRRORV6"
+ 4) "true"
+ 5) "ACL_ACTIONS|INGRESS"
+ 6) "PACKET_ACTION,REDIRECT_ACTION,MIRROR_INGRESS_ACTION"
+ 7) "ACL_ACTIONS|EGRESS"
+ 8) "PACKET_ACTION,MIRROR_EGRESS_ACTION"
+ 9) "ACL_ACTION|PACKET_ACTION"
+1)  "DROP,FORWARD"
+```
+In the above example output, two everflow combinations are supported on the platform being checked:
+* INGRESS stage, MIRROR_INGRESS_ACTION
+* EGRESS stage, MIRROR_EGRESS_ACTION
+
+This test plan needs to be extended to cover both the existing everflow function and the newly added capabilities.
+
+#### Some existing areas not covered by the existing scripts
+
+##### ACL rule for matching "IN_PORTS"
+Now the SONiC ACL rules support matching "IN_PORTS". New ACL rule for matching "IN_PORTS" need to be added and covered.
+
+##### IPv6 everflow
+The existing scripts only covered IPv6. IPv6 is also supported by SONiC now. The scripts need to be extended to cover IPv6 everflow too. To cover IPv6:
+* Everflow ACL table of type "MIRRORV6" needs to be defined and loaded during testing.
+* Different stages (ingress & egress) also need to be covered.
+* Different ACL rule mirror actions also need to be covered.
+* New set of IPv6 ACL rules need to be defined and loaded during testing.
+* The PTF script needs to be extended to inject and monitor IPv6 packets.
+
+#### How to extend the testing
+
+To cover the new enhancements, the existing scripts need to be extended:
+* The existing structure, sub-tests and PTF scripts can be reused.
+* Add new sets of configurations.
+* We can refactor the run_test.yml to run the existing sub-tests in multiple iteration. Each iteration loads a different set of ACL table and ACL rules configuration.
+* Adjust initialization of variables used in testing if needed.
+* Add a new class in the existing PTF script to inject and monitor IPv6 packets.
+* Add a sub-test and add a new class in the PTF script to cover ACL rule matching "IN_PORTS".
+
+In a summary, the extended scripts need to do below work:
+1. Firstly the script need to get capability info of the DUT from hard coded resource, then check against the detected capabilities in DB. Fail the test if capabilities do not match.
+2. Create everflow ACL table with different stage setting. Load ACL rules with different action type.
+3. Run test cases in the existing scripts (8 test cases at the time of writing).
+4. Ensure that each combination of the supported ACL table stages and ACL rule action types are covered.
+
+Summary of the possible combinations:
+
+Combinations | ACL table stage: ingress | ACL table stage: egress
+-|-|-
+ACL Rule MIRROR_INGRESS_ACTION | [x] | N/A
+ACL Rule MIRROR_EGRESS_ACTION | [x] | [x]
+
+Totally there are 3 possible combinations. Not all the combinations are supported by all platforms. The actual combinations to be tested are determined the actual DUT platform.
```

Comment:
I think of two options:
1. Query SWITCH_CAPABILITY table and decide which test cases to run and which to skip
2. Make user define this in some .yml configuration file per platform/hwsku/asic vendor

I am Ok with both. Since correctness of SWITCH_CAPABILITY table is verified by VS test I think you can rely on that during system level testing.
I am also Ok with .yml with configuration per platform/hwsku/asic.

Which one would you choose?

Reply:
I prefer option 1.